x-pack/plugins/security_solution/server/lib/detection_engine/rule_types/threat_marker/scan.ts

```diff
@@ -0,0 +1,230 @@
+  index: string[],
+  query: QueryDslQueryContainer,
+  indicator: string,
+  timestamp: number
+) =>
+  es.updateByQuery({
+    index,
+    query,
+    script: {
+      lang: 'painless',
+      source: `ctx._source["${THREAT_DETECTION_TIMESTAMP_FIELD}"] = params.timestamp; ctx._source["${THREAT_DETECTION_INDICATOR_FIELD}"] = params.indicator`,
+      params: {
+        timestamp,
+        indicator,
+      },
+    },
+    conflicts: 'proceed',
+    refresh: false,
+  });
+
+export const getDocuments = async <T = unknown>(
+  es: Client,
+  pit: string,
+  query?: QueryDslQueryContainer,
+  after?: any
+): Promise<Array<SearchHit<T>>> => {
+  const {
+    hits: { hits },
+  } = await es.search<T>({
+    pit: {
+      id: `${pit}`,
+      keep_alive: '1m',
+    },
+    size: 1000,
+    sort: ['@timestamp'],
+    ...(query ? { query } : {}),
+    ...(after ? { search_after: after } : {}),
+  });
+
+  return hits;
+};
+
+export const countDocuments = async (client: Client, index: string[]) =>
+  (await client.count({ index })).count;
+
+async function* documentGenerator<T>(
+  client: Client,
+  index: string[],
+  query?: QueryDslQueryContainer
+) {
+  let after: SortResults | undefined;
+
+  const pit: OpenPointInTimeResponse['id'] = (
+    await client.openPointInTime({
+      index,
+      keep_alive: '1m',
+    })
+  ).id;
+
+  while (true) {
+    const docs = await getDocuments<T>(client, pit, query, after);
+
+    if (!docs.length) {
+      break;
+    }
+
+    after = docs[docs.length - 1].sort;
+
+    yield docs;
+  }
+}
+
+export const scan = async (
+  { client, log }: { client: Client; log: (message: string) => void },
+  {
+    threatIndex,
+    eventsIndex,
+    concurrency,
+    verbose,
+  }: {
+    threatIndex: string[];
+    eventsIndex: string[];
+    concurrency: number;
+    verbose: boolean;
+  }
+) => {
+  const verboseLog = (message: string) => {
+    if (!verbose) {
+      return;
+    }
+
+    log(message);
+  };
+
+  log('update input indices mapping');
+
+  await updateMapping(client, eventsIndex);
+
+  log('starting scan');
+
+  const total = await countDocuments(client, threatIndex);
+  let progress = 0;
+
+  const start = Date.now();
+
+  for await (const threats of documentGenerator<Threat>(client, threatIndex)) {
+    await async.eachLimit(threats, concurrency, async ({ _source: threat, _id: threatId }) => {
+      progress++;
+
+      verboseLog(`processing threat ${threatId} (${progress}/${total})`);
+
+      if (!threat) {
+        verboseLog(`source is missing`);
+        return;
+      }
+
+      const shouldClause = shouldClauseForThreat(threat);
+
+      if (!shouldClause) {
+        verboseLog(
+          `skipping threat as no clauses are defined for its type: ${threat['threat.indicator.type']}`
+        );
+        return;
+      }
+
+      const query: QueryDslQueryContainer = {
+        bool: {
+          // This prevents processing events that were already checked. That means, after initial "scan",
```

Comment:
I think it's nice optimization but we need to check if it reflects reality. I can imagine cases when a source event matches multiple threats. Eg. if it has an URL, domain, and IP and all these were ingested as IoCs from some feeds. Though I can't say for sure if it's smth we need to care about. If it happens very rarely we can ignore this case as the benefit of keeping it simple might be much bigger

Reply:
Sure:)